// @strudel/midi/dist/index.mjs
import * as ce from "webmidi";
import { Note as ae } from "webmidi";
import { Pattern as J, isPattern as Q, logger as v, getControlName as X, noteToMidi as de, getIsStarted as ue, getCps as fe, getPattern as Z, getTriggerFunc as T, getTime as le, TimeSpan as W, Hap as V, ref as F, reify as me } from "@strudel/core";
import { getAudioContext as he } from "@strudel/webaudio";
let j;
const be = () => (j = new AudioContext(), j), ee = () => j || be();
let we = (e) => console.log(e);
const ge = (...e) => we(...e), ye = (e, n, t) => Math.min(Math.max(e, n), t);
function pe(e) {
  const n = ee().createGain();
  return n.gain.value = e, n;
}
function b(e, n, t = ee()) {
  const o = t.currentTime;
  Me(t, e, o, n);
}
function Me(e, n, t, o) {
  const r = new ConstantSourceNode(e), i = pe(0);
  return i.connect(e.destination), r.connect(i), xe(r, () => {
    G(i), G(r), n();
  }), r.start(t), r.stop(o), r;
}
const ne = (e) => e / (1 + e), ve = (e, n) => (e % n + n) % n, Ae = (e, n) => (1 + n) * e / (1 + n * Math.abs(e)), A = (e, n) => Math.tanh(e * (1 + n)), Ee = (e, n) => ye((1 + n) * e, -1, 1), te = (e, n) => {
  let t = (1 + 0.5 * n) * e;
  const o = ve(t + 1, 4);
  return 1 - Math.abs(o - 2);
}, Ce = (e, n) => Math.sin(Math.PI / 2 * te(e, n)), Ne = (e, n) => {
  const t = ne(Math.log1p(n)), o = (e - t / 3 * e * e * e) / (1 - t / 3);
  return A(o, n);
}, oe = (e, n, t = !1) => {
  const o = 1 + 2 * n, i = 0.07 * ne(Math.log1p(n)), s = A(e + i, 2 * n), c = A(t ? i : -e + i, 2 * n), m = s - c, w = 1 / Math.cosh(o * i), h = w * w, u = Math.max(1e-8, (t ? 1 : 2) * o * h);
  return A(m / u, n);
}, $e = (e, n) => oe(e, n, !0), _e = (e, n) => {
  const t = 10 * Math.log1p(n);
  let o = 1, r = e, i, s = 0;
  for (let c = 1; c < 64; c++) {
    if (c < 2) {
      s += c == 0 ? o : r;
      continue;
    }
    i = 2 * e * o - r, r = o, o = i, c % 2 === 0 && (s += Math.min(1.3 * t / c, 2) * i);
  }
  return A(s, t / 20);
}, Ie = {
  scurve: Ae,
  soft: A,
  hard: Ee,
  cubic: Ne,
  diode: oe,
  asym: $e,
  fold: te,
  sinefold: Ce,
  chebyshev: _e
};
Object.freeze(Object.keys(Ie));
const xe = (e, n) => {
  const t = n;
  e.onended = function() {
    t(), this.onended = null;
  };
}, G = (e) => {
  if (e != null) {
    if (!(e instanceof AudioNode))
      throw new Error("releaseAudioNode can only release an AudioNode");
    if (e.disconnect(), e instanceof AudioScheduledSourceNode) {
      process.env.NODE_ENV === "development" && e.onended && e.onended.name !== "cleanup" && ge(
        "[superdough] Deprecation warning: it seems your code path is setting 'node.onended = callback' instead of using the onceEnded helper"
      );
      try {
        e.stop();
      } catch {
        e.start(e.context.currentTime + 5), e.stop();
      }
    }
    e instanceof AudioWorkletNode && e.parameters.get("end")?.setValueAtTime(0, 0);
  }
}, { WebMidi: d } = ce;
function Se() {
  return typeof navigator.requestMIDIAccess == "function";
}
function $(e) {
  return e.map((n) => `'${n.name}'`).join(" | ");
}
function re(e = {}) {
  const { onReady: n, onConnected: t, onDisconnected: o, onEnabled: r } = e;
  if (!d.enabled) {
    if (!Se())
      throw new Error("Your Browser does not support WebMIDI.");
    return d.addListener("connected", () => {
      t?.(d);
    }), d.addListener("enabled", () => {
      r?.(d);
    }), d.addListener("disconnected", (i) => {
      o?.(d, i);
    }), new Promise((i, s) => {
      if (d.enabled) {
        i(d);
        return;
      }
      d.enable(
        (c) => {
          c && s(c), n?.(d), i(d);
        },
        { sysex: !0 }
      );
    });
  }
}
function H(e, n) {
  if (!n.length)
    throw new Error("🔌 No MIDI devices found. Connect a device or enable IAC Driver.");
  if (typeof e == "number")
    return n[e];
  const t = (i) => n.find((s) => s.name.includes(i));
  if (typeof e == "string")
    return t(e);
  const o = t("IAC"), r = o ?? n[0];
  if (!r)
    throw new Error(
      `🔌 MIDI device '${r || ""}' not found. Use one of ${$(n)}`
    );
  return o ?? n[0];
}
typeof window < "u" && window.addEventListener("message", (e) => {
  d?.enabled && e.data === "strudel-stop" && d.outputs.forEach((n) => n.sendStop());
});
const N = /* @__PURE__ */ new Map();
function ie(e) {
  return Object.fromEntries(
    Object.entries(e).map(([n, t]) => (typeof t == "number" && (t = { ccn: t }), [X(n), t]))
  );
}
function Pe(e, n = "") {
  if (!e.startsWith("github:"))
    throw new Error('expected "github:" at the start of pseudoUrl');
  let [t, o] = e.split("github:");
  return o = o.endsWith("/") ? o.slice(0, -1) : o, o.split("/").length === 2 && (o += "/main"), `https://raw.githubusercontent.com/${o}/${n}`;
}
function Ue(e) {
  N.set("default", ie(e));
}
let B = {};
async function Re(e) {
  typeof e == "string" && (e.startsWith("github:") && (e = Pe(e, "midimap.json")), B[e] || (B[e] = fetch(e).then((n) => n.json())), e = await B[e]), typeof e == "object" && Object.entries(e).forEach(([n, t]) => N.set(n, ie(t)));
}
const Ve = /* @__PURE__ */ new Map();
function je(e = 0, n = 0, t = 1, o = 1) {
  if (n === t)
    throw new Error("min and max cannot be the same value");
  let r = (e - n) / (t - n);
  return r = Math.min(1, Math.max(0, r)), Math.pow(r, o);
}
function ke(e, n) {
  return Object.keys(n).filter((t) => !!e[X(t)]).map((t) => {
    const { ccn: o, min: r = 0, max: i = 1, exp: s = 1 } = e[t], c = je(n[t], r, i, s);
    return { ccn: o, ccv: c };
  });
}
function q(e, n, t, o, r) {
  if (typeof n != "number" || n < 0 || n > 1)
    throw new Error("expected ccv to be a number between 0 and 1");
  if (!["string", "number"].includes(typeof e))
    throw new Error("expected ccn to be a number or a string");
  const i = Math.round(n * 127);
  b(() => {
    t.sendControlChange(e, i, o);
  }, r);
}
function K(e, n, t, o) {
  if (typeof e != "number" || e < 0 || e > 127)
    throw new Error("expected progNum (program change) to be a number between 0 and 127");
  b(() => {
    n.sendProgramChange(e, t);
  }, o);
}
function Y(e, n, t, o) {
  if (Array.isArray(e)) {
    if (!e.every((r) => Number.isInteger(r) && r >= 0 && r <= 255))
      throw new Error("all sysexid bytes must be integers between 0 and 255");
  } else if (!Number.isInteger(e) || e < 0 || e > 255)
    throw new Error("A:sysexid must be an number between 0 and 255 or an array of such integers");
  if (!Array.isArray(n))
    throw new Error("expected sysex to be an array of numbers (0-255)");
  if (!n.every((r) => Number.isInteger(r) && r >= 0 && r <= 255))
    throw new Error("all sysex bytes must be integers between 0 and 255");
  b(() => {
    t.sendSysex(e, n);
  }, o);
}
function De(e, n, t, o, r) {
  if (Array.isArray(e)) {
    if (!e.every((i) => Number.isInteger(i) && i >= 0 && i <= 255))
      throw new Error("all nrpnn bytes must be integers between 0 and 255");
  } else if (!Number.isInteger(n) || n < 0 || n > 255)
    throw new Error("A:sysexid must be an number between 0 and 255 or an array of such integers");
  b(() => {
    t.sendNRPN(e, n, o);
  }, r);
}
function Oe(e, n, t, o) {
  if (typeof e != "number" || e < -1 || e > 1)
    throw new Error("expected midibend to be a number between -1 and 1");
  b(() => {
    n.sendPitchBend(e, t);
  }, o);
}
function Le(e, n, t, o) {
  if (typeof e != "number" || e < 0 || e > 1)
    throw new Error("expected miditouch to be a number between 0 and 1");
  b(() => {
    n.sendChannelAftertouch(e, t);
  }, o);
}
function We(e, n, t, o, r, i) {
  if (e == null || e === "")
    throw new Error("note cannot be null or empty");
  if (n != null && (typeof n != "number" || n < 0 || n > 1))
    throw new Error("velocity must be a number between 0 and 1");
  if (t != null && (typeof t != "number" || t < 0))
    throw new Error("duration must be a positive number");
  const s = typeof e == "number" ? e : de(e), c = new ae(s, { attack: n, duration: t });
  b(() => {
    o.playNote(c, r);
  }, i);
}
J.prototype.midi = function(e, n = {}) {
  if (Q(e))
    throw new Error(
      `.midi does not accept Pattern input for midiport. Make sure to pass device name with single quotes. Example: .midi('${d.outputs?.[0]?.name || "IAC Driver Bus 1"}')`
    );
  if (typeof e == "object") {
    const { port: o, isController: r = !1, ...i } = e;
    n = {
      isController: r,
      ...i,
      ...n
      // Keep any options passed separately
    }, e = o;
  }
  let t = {
    // Default configuration values
    isController: !1,
    // Disable sending notes for midi controllers
    noteOffsetMs: 10,
    // Default note-off offset to prevent glitching in ms
    midichannel: 1,
    // Default MIDI channel
    velocity: 0.9,
    // Default velocity
    gain: 1,
    // Default gain
    midimap: "default",
    // Default MIDI map
    midiport: e,
    // Store the port in the config
    ...n
    // Override defaults with provided options
  };
  return re({
    onEnabled: ({ outputs: o }) => {
      const r = H(t.midiport, o), i = o.filter((s) => s.name !== r.name);
      v(
        `Midi enabled! Using "${r.name}". ${i?.length ? `Also available: ${$(i)}` : ""}`
      );
    },
    onDisconnected: ({ outputs: o }) => v(`Midi device disconnected! Available: ${$(o)}`)
  }), this.onTrigger((o, r, i, s) => {
    if (!d.enabled) {
      v("Midi not enabled");
      return;
    }
    o.ensureObjectValue();
    let {
      note: c,
      nrpnn: m,
      nrpv: w,
      ccn: h,
      ccv: u,
      midichan: l = t.midichannel,
      midicmd: a,
      midibend: g,
      miditouch: E,
      polyTouch: k,
      gain: _ = t.gain,
      velocity: y = t.velocity,
      progNum: p,
      sysexid: D,
      sysexdata: U,
      midimap: I = t.midimap,
      midiport: R = t.midiport
    } = o.value;
    const f = H(R, d.outputs);
    if (!f) {
      v(
        `[midi] midiport "${R}" not found! available: ${d.outputs.map((C) => `'${C.name}'`).join(", ")}`
      );
      return;
    }
    if (y = _ * y, N.has(I) ? ke(N.get(I), o.value).forEach(({ ccn: O, ccv: L }) => q(O, L, f, l, s)) : I !== "default" && v(`[midi] midimap "${I}" not found! Available maps: ${[...N.keys()].join(", ")}`), c !== void 0 && !t.isController) {
      const C = o.duration.valueOf() / i * 1e3 - t.noteOffsetMs;
      We(c, y, C, f, l, s);
    }
    if (p !== void 0 && K(p, f, l, s), D !== void 0 && U !== void 0 && Y(D, U, f, s), u !== void 0 && h !== void 0 && q(h, u, f, l, s), m !== void 0 && w !== void 0 && De(m, w, f, l, s), g !== void 0 && Oe(g, f, l, s), E !== void 0 && Le(E, f, l, s), o.whole.begin + 0 === 0 && b(() => {
      f.sendStart();
    }, s), ["clock", "midiClock"].includes(a))
      b(() => {
        f.sendClock();
      }, s);
    else if (["start"].includes(a))
      b(() => {
        f.sendStart();
      }, s);
    else if (["stop"].includes(a))
      b(() => {
        f.sendStop();
      }, s);
    else if (["continue"].includes(a))
      b(() => {
        f.sendContinue();
      }, s);
    else if (Array.isArray(a)) {
      if (a[0] === "progNum")
        K(a[1], f, l, s);
      else if (a[0] === "cc")
        a.length === 2 && q(a[0], a[1] / 127, f, l, s);
      else if (a[0] === "sysex" && a.length === 3) {
        const [C, O, L] = a;
        Y(O, L, f, s);
      }
    }
  });
};
async function se(e) {
  if (Q(e))
    throw new Error(
      `[midi] Midi input cannot be a pattern. Make sure to pass device name with single quotes. Example: midin('${d.outputs?.[0]?.name || "IAC Driver Bus 1"}')`
    );
  const n = await re(), t = H(e, d.inputs);
  if (!t)
    throw new Error(
      `[midi] Midi device "${e}" not found.. connected devices: ${$(d.inputs)}`
    );
  if (n) {
    const o = d.inputs.filter((r) => r.name !== t.name);
    v(
      `[midi] Midi enabled! Using "${t.name}". ${o?.length ? `Also available: ${$(o)}` : ""}`
    );
  }
  return t;
}
let x = {};
const z = {}, S = {};
async function Fe(e) {
  const n = await se(e);
  z[e] ??= {}, S[e] ??= {};
  const t = (o, r) => r !== void 0 ? F(() => S[e][o]?.[r] || 0) : F(() => z[e][o] || 0);
  return x[e] && n.removeListener("midimessage", x[e]), x[e] = (o) => {
    const [r, i] = o.dataBytes, s = o.message.channel, c = i / 127;
    S[e][r] ??= {}, S[e][r][s] = c, z[e][r] = c;
  }, n.addListener("midimessage", x[e]), t;
}
const M = {}, P = {};
function Be(e, n, t, o) {
  const r = Z(), i = T();
  if (!r || !i)
    return !1;
  const s = t + o, c = 1e-6, w = r.queryArc(s - c, s + c, { _cps: n }).filter((u) => u.value?.midikey?.startsWith(`${e}_`)), h = he().currentTime;
  return w.length ? (w.forEach((u) => {
    if (!u.hasOnset())
      return;
    const l = h + (u.whole.begin - t) / n, a = u.duration / n;
    i(u, l - h, a, n, l);
  }), !0) : !1;
}
async function Ge(e) {
  const n = await se(e);
  return M[e] || (M[e] = []), P[e] && n.removeListener("midimessage", P[e]), P[e] = (o) => {
    const { dataBytes: r, message: i } = o, s = i.command === 9;
    let c = i.command === 8;
    if (!ue() || !s && !c)
      return;
    const [h, u] = r;
    c ||= s && u === 0;
    const l = `${e}_${h}`, a = fe() ?? 0.5, g = !!(Z() && T()), E = g ? 0.01 : 0.06, k = le(), _ = k + E * a, y = new W(_, _);
    let p = { midikey: l };
    c || (p = { ...p, note: Math.round(h), velocity: u / 127 }, M[e].push(new V(y, y, p, {})), !c && g && Be(e, a, k, E * a) && (M[e] = []));
  }, n.addListener("midimessage", P[e]), (o = 0.5) => {
    const r = me(o), i = (s) => {
      const c = M[e].flatMap((m) => r.query(s.setSpan(m.wholeOrPart())).map((h) => {
        const u = h.value ?? 0.5, l = new W(m.whole.begin, m.whole.begin.add(u)), a = new W(m.part.begin, m.part.begin.add(u)), g = m.combineContext(h);
        return new V(l, a, m.value, g);
      }));
      return s.controls.cyclist && (M[e] = []), c;
    };
    return new J(i);
  };
}
export {
  d as WebMidi,
  Ue as defaultmidimap,
  re as enableWebMidi,
  N as midicontrolMap,
  Ge as midikeys,
  Re as midimaps,
  Fe as midin,
  Ve as midisoundMap
};
